jupyterhub/auth.py

```diff
@@ -746,19 +749,19 @@ def authenticate(self, handler, data):
         return username
 
 
-    @run_on_executor
-    def pre_spawn_start(self, user, spawner):
+    def pre_exec_fn(self, user, spawner):
         """Open PAM session for user if so configured"""
         if not self.open_sessions:
             return
         try:
             pamela.open_session(user.name, service=self.service, encoding=self.encoding)
         except pamela.PAMError as e:
             self.log.warning("Failed to open PAM session for %s: %s", user.name, e)
-            self.log.warning("Disabling PAM sessions from now on.")
-            self.open_sessions = False
+            # this will be ineffective in the child process
+            # self.log.warning("Disabling PAM sessions from now on.")
+            # self.open_sessions = False
 
-    @run_on_executor
+    #@run_on_executor
     def post_spawn_stop(self, user, spawner):
```

Comment:
As I understand it, to fix the issue, the close session also needs to happen in the subprocess as well, not just the open. How do we accomplish that? There isn't an equivalent to preexec_fn for the end of a process.

Do we instead need to create a two-layer process, where the parent does:

```python
pamela.open_session()
p = start_real_subprocess(args, env)
wait(p)
pamela.close_session()
```

Reply:
This is believe this is how `sshd` accomplishes it. It has process that watches the shell process.

I think creating a `PAMAuthenticator.post_exec_stop()` function and moving the PAM session logic into it would be best and making the two-layer process look like this:
```
if self.auth:
   self.auth.pre_exec_fn(self.user, self)
Popen.run(args, env) # will wait for process to end on its own
if self.auth:
   self.auth.post_exec_stop(self.user, self)
```

This also has the advantage of allowing me to avoid the `@run_on_executor` statement that has caused issue with one check, because the PAM session will already be occurring off of the main thread.

Reply:
It also avoid doing a lot with `pre_execfn` kwarg which the Python docs say is bad.